File: 064_MinimumPathSum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m, n = len(grid), len(grid[0])
        curr = grid[m - 1][:]
        for i in range(n - 1)[::-1]:
            curr[i] += curr[i + 1]
        
        for i in range(m - 1)[::-1]:
            curr[n - 1] += grid[i][n - 1]
            for j in range(n - 1)[::-1]:
                curr[j] = min(curr[j], curr[j + 1]) + grid[i][j]
        return curr[0]

#In-place, without using extra space. quite a smart one		
def minPathSum(self, grid):
    m = len(grid)
    n = len(grid[0])
    for i in range(1, n):
        grid[0][i] += grid[0][i-1]
    for i in range(1, m):
        grid[i][0] += grid[i-1][0]
    for i in range(1, m):
        for j in range(1, n):
            grid[i][j] += min(grid[i-1][j], grid[i][j-1])
    return grid[-1][-1]

File: 064_MinimumPathSum.cpp
//28.72%
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m(grid.size()), n(grid[0].size());
		vector<int> curr(grid[m - 1]);
		for(int i = n - 2; i >= 0; i--)
			curr[i] = curr[i + 1] + grid[m - 1][i];
		
		for(int i = m - 2; i >= 0; i--){
			curr[n - 1] += grid[i][n - 1];
			for(int j = n - 2; j >= 0; j--){
				curr[j] = min(curr[j], curr[j + 1]) + grid[i][j];
			}
		}
		return curr[0];
    }
};
//Suddenly, i realized from 'up left' to 'down right' problem equals to the problem from 'down right' to 'top left'..^_^ 
//Same principle like mine
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<int> cur(m, grid[0][0]);
        for (int i = 1; i < m; i++)
            cur[i] = cur[i - 1] + grid[i][0]; 
        for (int j = 1; j < n; j++) {
            cur[0] += grid[0][j]; 
            for (int i = 1; i < m; i++)
                cur[i] = min(cur[i - 1], cur[i]) + grid[i][j];
        }
        return cur[m - 1];
    }
};